### PR TITLE
Fix #17629  #query_ldap issues

### LINKED ISSUE #17629: Fix Broken Logic for 64 bit Systems when using enum_ad_users.rb
Whilst trying to fix the Railgun definitions in https://github.com/rapid7/metasploit-framework/pull/17562, we have been trying to use `post/windows/gather/enum_ad_users.rb` to do our testing. However we have honestly run into so many issues that are not Railgun specific with that module we thought it was about time we put them into a separate ticket since this module needs a good bit of reworking at the moment to work on anything outside of a limited set of payloads and host types.

Note that all testing requires one to use a payload type where extapi extension is not loaded, such as a Python Meterpreter payload, as otherwise the RailGun backend will not be used and some of these issues won't be hit.

Also in case you wondering why we just don't use the Net::LDAP library, which is something I raised as a concern, apparently we want to still support authenticating as the current user so we are using https://learn.microsoft.com/en-us/windows/win32/api/winldap/nf-winldap-ldap_bind_sa with `LDAP_AUTH_NEGOTIATE` which allows us to do this: `To log in as the current user, set the dn and cred parameters to NULL.`.

Here are some of the issues that were encountered:
1. https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/windows/ldap.rb#L267 is trying to read what appears to be a https://learn.microsoft.com/en-us/windows/win32/api/winldap/ns-winldap-ldapmessage structure however it fails to account for the fact that some elements may be 64 bit pointers, so this will fail on x64 systems.
2. We appear to be doing some operations on https://learn.microsoft.com/en-us/windows/win32/api/winldap/ns-winldap-berelement however the MSDN documentation lists this structure as being opaque. Despite this, we appear to be trying to play around with this at https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/windows/ldap.rb#L279-L286 for reasons unknown.
3. The logic at https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/windows/ldap.rb#L282 is working on architectures incorrectly. If we use a Python session the arch will be ARCH_PYTHON which won't match things. We should be using the `native_arch` attribute like https://github.com/rapid7/metasploit-framework/blob/e26ab2c202fe6bc6b90e383bee6e20233966c6e5/lib/rex/post/meterpreter/extensions/stdapi/railgun/multicall.rb#L77 does. 
4. Currently https://github.com/rapid7/metasploit-framework/blob/86fc6172591616dccedbe9bb4c8abbcc27fdcd4f/lib/msf/core/post/windows/ldap.rb#L130 is broken when called from a 64 bit Meterpreter session without the `extapi` extension loaded.
 

## Steps to reproduce

How'd you do it?

1. Get a session on a x64 system with a Python Meterpreter shell.
2. Try run the `post/windows/gather/enum_ad_users` module, set the `DOMAIN` value to your domain, and the `SESSION` to your Python session.
3. You should encounter some of the issues noted above. Note that you may need a debugging session to force some of the code paths so you can test edge cases and see that some of the logic might not be working.

## Were you following a specific guide/tutorial or reading documentation?

Found whilst doing testing for https://github.com/rapid7/metasploit-framework/pull/17562

## Expected behavior

The `enum_ad_users.rb` module should work on x64 systems even if the `extapi` extension is not loaded.

## Current behavior

It will fail in a variety of fashions listed above, plus potentially extra we haven't encountered yet.

### Metasploit version

6.3.1-dev-e26ab2c202

### PR DESCRIPTION
**Pull request by zeroSteiner**
This fixes the `#query_ldap` post API method. The original would parse the opaque BER data structures manually which lead to issues due to the assumption that everything was 32-bit. Now that #17562 has been landed the native Windows API methods can be used regardless of if the target is 32-bit or 64-bit. 

The old implementation would obtain the `LDAPMessage` structure and then process it manually, going as far as decoding the BER data itself. I can't find any documentation on the structure of the `LDAPMessage` so it would seem that Microsoft doesn't intend for it to be accessed.

The new implementation takes the following steps:

1. Calls [`ldap_first_entry`](https://learn.microsoft.com/en-us/windows/win32/api/winldap/nf-winldap-ldap_first_entry) to obtain the first `LDAPMessage`
2. For each field that was requested, it calls [`ldap_get_values`](https://learn.microsoft.com/en-us/windows/win32/api/winldap/nf-winldap-ldap_get_values)
3. Counts the number of values that were returned
4. Reads that array of pointers with `#memread`, unpacking them based on the architecture
5. Reads the string data from each and joins it all together using a comma
6. Frees the values
7. Iterates through the next entry, going back to step 2
8. Frees the LDAPMessage

Step 5 is incorrect, we should really be returning an array of strings but the original implementation also just concatenated everything together with a comma. The underlying issue is that LDAP values can be nil a single value or an array of multiple values. I left this as is because modules using this API are probably accustomed to the value being a string.

For testing, use a Python Meterpreter because it supports railgun and the existing logic prefers the extapi for the Windows Meterpreter.

- [x] Start a Python Meterpreter session on a domain controller (this is the easiest to get the LDAP connection to just work)
- [x] Run the `post/windows/gather/enum_ad_users` module and see it work
- [x] Run the `post/windows/gather/enum_ad_computers` module and see it work

## Demo (Fixed)

```
meterpreter > sysinfo
Computer        : DC
OS              : Windows 2016 (Build 17763)
Architecture    : x64
System Language : en_US
Meterpreter     : python/windows
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > run post/windows/gather/enum_ad_users

[!] SESSION may not be compatible with this module:
[!]  * unloadable Meterpreter extension: extapi
[*] Querying default naming context
Domain Users
============

 sAMAccountName        name                              userPrincipalName                 userAccountControl  lockoutTime  mail                                    primarygroupid  description
 --------------        ----                              -----------------                 ------------------  -----------  ----                                    --------------  -----------
 Administrator         Administrator                                                       66048                                                                    513             Built-in account for administering the computer/domain
 Guest                 Guest                                                               66082                                                                    514             Built-in account for guest access to the computer/domain
 smcintyre             smcintyre                         smcintyre@msflab.local            66048                            smcintyre@msflab.local                  513
 sshd                  sshd                                                                66048                                                                    513
 krbtgt                krbtgt                                                              514                                                                      513             Key Distribution Center Service Account
 aliddle               Alice Liddle                      aliddle@msflab.local              66048                            aliddle@msflab.local                    513
 $MI1000-P4PGUIV8EHR1  Exchange Online-ApplicationAccou  Exchange_Online-ApplicationAccou  546                                                                      513
                       nt                                nt@msflab.local
 SM_05950467e3ea46ed8  SystemMailbox{1f05a927-1c97-4df7  SystemMailbox{1f05a927-1c97-4df7  514                              SystemMailbox{1f05a927-1c97-4df7-ad98-  513
                       -ad98-cab0bb4191db}               -ad98-cab0bb4191db}@msflab.local                                   cab0bb4191db}@msflab.local
 SM_274b5effbbf3459f8  SystemMailbox{bb558c35-97f1-4cb9  SystemMailbox{bb558c35-97f1-4cb9  514                              SystemMailbox{bb558c35-97f1-4cb9-8ff7-  513
                       -8ff7-d53741dc928c}               -8ff7-d53741dc928c}@msflab.local                                   d53741dc928c}@msflab.local
 SM_1ea45a0f5e274d5c9  SystemMailbox{e0dc1c29-89c3-4034  SystemMailbox{e0dc1c29-89c3-4034  514                              SystemMailbox{e0dc1c29-89c3-4034-b678-  513
                       -b678-e6c29d823ed9}               -b678-e6c29d823ed9}@msflab.local                                   e6c29d823ed9}@msflab.local
 SM_1ec2821cf0ea4ec19  DiscoverySearchMailbox {D919BA05  DiscoverySearchMailbox {D919BA05  514                              DiscoverySearchMailbox{D919BA05-46A6-4  513
                       -46A6-415f-80AD-7E09334BB852}     -46A6-415f-80AD-7E09334BB852}@ms                                   15f-80AD-7E09334BB852}@msflab.local
                                                         flab.local
 SM_164251cf2e97458d8  Migration.8f3e7716-2011-43e4-96b  Migration.8f3e7716-2011-43e4-96b  514                              Migration.8f3e7716-2011-43e4-96b1-aba6  513
                       1-aba62d229136                    1-aba62d229136@msflab.local                                        2d229136@msflab.local
 SM_d1601338536f49488  FederatedEmail.4c1f4d8b-8179-414  FederatedEmail.4c1f4d8b-8179-414  514                              FederatedEmail.4c1f4d8b-8179-4148-93bf  513
                       8-93bf-00a95fa1e042               8-93bf-00a95fa1e042@msflab.local                                   -00a95fa1e042@msflab.local
 SM_7d14c5fb0a244bb0b  SystemMailbox{D0E409A0-AF9B-4720  SystemMailbox{D0E409A0-AF9B-4720  514                              SystemMailbox{D0E409A0-AF9B-4720-92FE-  513
                       -92FE-AAC869B0D201}               -92FE-AAC869B0D201}@msflab.local                                   AAC869B0D201}@msflab.local
 SM_9fbfdc9f9da34155b  SystemMailbox{2CE34405-31BE-455D  SystemMailbox{2CE34405-31BE-455D  514                              SystemMailbox{2CE34405-31BE-455D-89D7-  513
                       -89D7-A7C7DA7A0DAA}               -89D7-A7C7DA7A0DAA}@msflab.local                                   A7C7DA7A0DAA}@msflab.local
 SM_c6906a0fc1764b5db  SystemMailbox{8cc370d3-822a-4ab8  SystemMailbox{8cc370d3-822a-4ab8  514                              SystemMailbox{8cc370d3-822a-4ab8-a926-  513
                       -a926-bb94bd0641a9}               -a926-bb94bd0641a9}@msflab.local                                   bb94bd0641a9}@msflab.local
 HealthMailbox21b1853  HealthMailbox21b1853745514961993  HealthMailbox21b1853745514961993  66048                            HealthMailbox21b185374551496199303bc32  513
                       03bc32fee6069                     03bc32fee6069@msflab.local                                         fee6069@msflab.local
 HealthMailboxd46a71f  HealthMailboxd46a71f1e7ae4e4a896  HealthMailboxd46a71f1e7ae4e4a896  66048                            HealthMailboxd46a71f1e7ae4e4a896d9485a  513
                       d9485a0a4bf4c                     d9485a0a4bf4c@msflab.local                                         0a4bf4c@msflab.local
 HealthMailboxf8d9ea7  HealthMailboxf8d9ea752c084e7293a  HealthMailboxf8d9ea752c084e7293a  66048                            HealthMailboxf8d9ea752c084e7293ad6a5d2  513
                       d6a5d20f6f554                     d6a5d20f6f554@msflab.local                                         0f6f554@msflab.local
 HealthMailbox94d093a  HealthMailbox94d093a0f05e48ad811  HealthMailbox94d093a0f05e48ad811  66048                            HealthMailbox94d093a0f05e48ad81124c19b  513
                       24c19bb6fa748                     24c19bb6fa748@msflab.local                                         b6fa748@msflab.local
 HealthMailboxa9a11e7  HealthMailboxa9a11e74426c436aa75  HealthMailboxa9a11e74426c436aa75  66048                            HealthMailboxa9a11e74426c436aa759c6687  513
                       9c6687cc32754                     9c6687cc32754@msflab.local                                         cc32754@msflab.local
 HealthMailbox8d864d2  HealthMailbox8d864d2afc2749feaef  HealthMailbox8d864d2afc2749feaef  66048                            HealthMailbox8d864d2afc2749feaef57bc62  513
                       57bc62d240431                     57bc62d240431@msflab.local                                         d240431@msflab.local
 HealthMailbox9a524ee  HealthMailbox9a524ee043dd4124ad9  HealthMailbox9a524ee043dd4124ad9  66048                            HealthMailbox9a524ee043dd4124ad9ed993c  513
                       ed993c5438561                     ed993c5438561@msflab.local                                         5438561@msflab.local
 HealthMailbox8e506cb  HealthMailbox8e506cb9bed24339bef  HealthMailbox8e506cb9bed24339bef  66048                            HealthMailbox8e506cb9bed24339befd24886  513
                       d248865cb00a4                     d248865cb00a4@msflab.local                                         5cb00a4@msflab.local
 HealthMailboxc1b73ec  HealthMailboxc1b73eca8fbb40ad85c  HealthMailboxc1b73eca8fbb40ad85c  66048                            HealthMailboxc1b73eca8fbb40ad85cec65b4  513
                       ec65b466c864c                     ec65b466c864c@msflab.local                                         66c864c@msflab.local
 HealthMailboxf9f877e  HealthMailboxf9f877e335d24143828  HealthMailboxf9f877e335d24143828  66048                            HealthMailboxf9f877e335d24143828deb0d0  513
                       deb0d0fff7634                     deb0d0fff7634@msflab.local                                         fff7634@msflab.local
 HealthMailboxa56bf05  HealthMailboxa56bf053817b43cca68  HealthMailboxa56bf053817b43cca68  66048                            HealthMailboxa56bf053817b43cca68cfb756  513
                       cfb756f4db097                     cfb756f4db097@msflab.local                                         f4db097@msflab.local
 mhatter               Maddilyn Hatter                   mhatter@msflab.local              16843264                                                                 513

meterpreter > run post/windows/gather/enum_ad_computers

[!] SESSION may not be compatible with this module:
[!]  * unloadable Meterpreter extension: extapi
[*] Querying default naming context
Domain Computers
================

 dNSHostName                distinguishedName                                description  operatingSystem               operatingSystemServicePack
 -----------                -----------------                                -----------  ---------------               --------------------------
 DC.msflab.local            CN=DC,OU=Domain Controllers,DC=msflab,DC=local                Windows Server 2019 Standard
 EXCHANGE2019.msflab.local  CN=EXCHANGE2019,CN=Computers,DC=msflab,DC=local               Windows Server 2019 Standard
 MSSQL.msflab.local         CN=MSSQL,CN=Computers,DC=msflab,DC=local                      Windows Server 2019 Standard

meterpreter > 
```

## Demo (Original issue)

```
meterpreter > sysinfo
Computer        : DC
OS              : Windows 2016 (Build 17763)
Architecture    : x64
System Language : en_US
Meterpreter     : python/windows
meterpreter > run post/windows/gather/enum_ad_users

[!] SESSION may not be compatible with this module:
[!]  * unloadable Meterpreter extension: extapi
[-] Unable to find the domain to query.
meterpreter > run post/windows/gather/enum_ad_computers

[!] SESSION may not be compatible with this module:
[!]  * unloadable Meterpreter extension: extapi
[-] Unable to find the domain to query.
meterpreter > 
```

### CODE REVIEW

File: modules/post/windows/gather/enum_ad_users.rb [hunk 1]
```diff
@@ -157,7 +157,7 @@ def query_filter
     inner_filter << '(!(lockoutTime>=1))' if datastore['EXCLUDE_LOCKED']
     inner_filter << '(!(userAccountControl:1.2.840.113556.1.4.803:=2))' if datastore['EXCLUDE_DISABLED']
     inner_filter << "(memberof:1.2.840.113556.1.4.1941:=#{datastore['GROUP_MEMBER']})" if datastore['GROUP_MEMBER']
-    inner_filter << "(#{datastore['FILTER']})" if datastore['FILTER'] != ''
+    inner_filter << "(#{datastore['FILTER']})" unless datastore['FILTER'].blank?
```
Comment (zeroSteiner) [inline modules/post/windows/gather/enum_ad_users.rb:160]: If `datastore['FILTER']` was nil, this would cause `()` to be added to the filter which is invalid and that would cause the query to break.

**Comment by bwatters-r7** [comment]
Old and busted
```
msf6 post(windows/gather/enum_ad_users) > sessions -l

Active sessions
===============

  Id  Name  Type                        Information                                Connection
  --  ----  ----                        -----------                                ----------
  1         meterpreter x64/windows     NT AUTHORITY\SYSTEM @ APT_WIN2016X64       10.5.135.201:4567 -> 10.5.134.159:64506 (10.5.13
                                                                                   4.159)
  2         meterpreter python/windows  TESTDOMAIN\Administrator @ APT_Win2016x64  10.5.135.201:4585 -> 10.5.134.159:64520 (10.5.13
                                                                                   4.159)

msf6 post(windows/gather/enum_ad_users) > set session 2
session => 2
msf6 post(windows/gather/enum_ad_users) > run

[!] SESSION may not be compatible with this module:
[!]  * unloadable Meterpreter extension: extapi
[-] Unable to find the domain to query.
[*] Post module execution completed
msf6 post(windows/gather/enum_ad_users) > 

```

New and Improved
```
msf6 payload(python/meterpreter/reverse_tcp) > [*] Sending stage (24380 bytes) to 10.5.134.159
[*] Meterpreter session 1 opened (10.5.135.201:4585 -> 10.5.134.159:64537) at 2023-02-27 13:46:13 -0600

msf6 payload(python/meterpreter/reverse_tcp) > use post/windows/gather/enum_ad_users
msf6 post(windows/gather/enum_ad_users) > set session 1
session => 1
msf6 post(windows/gather/enum_ad_users) > run

[!] SESSION may not be compatible with this module:
[!]  * unloadable Meterpreter extension: extapi
[*] Querying default naming context
Domain Users
============

 sAMAccountName  name            userPrincipalN  userAccountCon  lockoutTime  mail  primarygroupid  description
                                 ame             trol
 --------------  ----            --------------  --------------  -----------  ----  --------------  -----------
 Administrator   Administrator                   66048                              513             Built-in account for administer
                                                                                                    ing the computer/domain
 Guest           Guest                           66082                              514             Built-in account for guest acce
                                                                                                    ss to the computer/domain
 DefaultAccount  DefaultAccount                  66082                              513             A user account managed by the s
                                                                                                    ystem.
 msfuser         msfuser                         66048                              513
 krbtgt          krbtgt                          514                                513             Key Distribution Center Service
                                                                                                     Account

[*] Post module execution completed
msf6 post(windows/gather/enum_ad_users) > 

```

**Comment by bwatters-r7** [comment]
# Release  Notes
This PR changes the behavior from manually parsing the `LDAPMessage` manually and instead relies on railgun calls to native Windows API methods to parse the structure.
